File: packages/csharp/PriorityQueue.cs
namespace csharp_packages;
using System;

public class MyPriorityQueue<T> where T : IComparable<T> {
    private readonly MinHeap<T> _heap = new();
    public void Enqueue(T item) => _heap.Insert(item);
    public T Dequeue() => _heap.ExtractMin();
    public bool IsEmpty => _heap.IsEmpty;
}

File: packages/zig/src/priorityqueue.zig
const std = @import("std");
const Heap = @import("heap.zig").Heap;
const Allocator = std.mem.Allocator;

pub fn PriorityQueue(comptime T: type, comptime compareFn: fn (a: T, b: T) bool) type {
    return struct {
        heap: Heap(T, compareFn),

        const Self = @This();

        pub fn init(allocator: Allocator) Self {
            return Self{ .heap = Heap(T, compareFn).init(allocator) };
        }

        pub fn deinit(self: *Self) void {
            self.heap.deinit();
        }

        pub fn enqueue(self: *Self, item: T) !void {
            try self.heap.insert(item);
        }

        pub fn dequeue(self: *Self) ?T {
            return self.heap.extract();
        }
    };
}